File: app/toes/xml_parser.py
import enum
import os.path

from app.toes.node import Node
from app.toes.text_node import TextNode
from app.toes.root_node import RootNode
from app.toes.comment_node import CommentNode
from app.toes.processing_node import ProcessingNode
from app.toes.directive_node import DirectiveNode
from app.toes.toes_exceptions import XMLParsingException

from app.utilities import positive_min


class STATES(enum.Enum):
    new_page = 0
    read_node_name = 8
    looking_for_attribute = 9
    looking_for_child_nodes = 18
    inside_script = 19


class XmlParsingInfo:
    def __init__(self, current_node: Node = None):
        self.i = 0
        self.state = STATES.new_page
        self.current_node: Node = current_node
        self.root_node: Node = current_node

    def move_index(self, step: int = 1):
        self.i += step


class XMLParser:

    def __init__(self, *args, path=None, template=None, base_path=None, **kwargs):
        if path is not None:
            if base_path is not None:
                path = os.path.join(base_path, path)
            with open(path, mode="r", encoding="utf-8") as text_file:
                self.text = text_file.read()
        elif template is not None:
            self.text = template

    def parse_file(self):
        if self.text is None:
            return "Error: empty text"
        if len(self.text) == 0:
            return self.text

        result = self.text
        parsing_info = XmlParsingInfo(current_node=RootNode())
        while parsing_info.i < len(result):
            if result[parsing_info.i] == "<":
                result, parsing_info = self.parse_starting_tag_character(text=result, parsing_info=parsing_info)
            elif result[parsing_info.i] == ">":
                result, parsing_info = self.parse_ending_tag_character(text=result, parsing_info=parsing_info)
            elif result[parsing_info.i].isspace():

                parsing_info.move_index()
            else:
                result, parsing_info = self.parse_character(text=result, parsing_info=parsing_info)

        return parsing_info.root_node

    def parse_starting_tag_character(self, text: str, parsing_info: XmlParsingInfo) -> (str, XmlParsingInfo):
        if text[parsing_info.i + 1] == " ":
            parsing_info.move_index()
        elif parsing_info.state == STATES.new_page:
            if text[parsing_info.i:].find("<?xml") == 0:
                parsing_info.move_index(len("<?xml"))
                parsing_info.state = STATES.looking_for_attribute
            elif text[parsing_info.i:].find("<!DOCTYPE") == 0:
                parsing_info.current_node.html = True
                parsing_info.current_node.attributes["doctype"] = text[len("<!DOCTYPE "):text.find(">")]
                parsing_info.state = STATES.looking_for_child_nodes
                parsing_info.move_index(text.find(">"))
            else:
                parsing_info = self.create_new_node(text, parsing_info)
                parsing_info.state = STATES.read_node_name
        elif parsing_info.state == STATES.looking_for_child_nodes:
            if text[parsing_info.i:].find("<![CDATA[") == 0:
                parsing_info.move_index(len("<![CDATA["))
                parsing_info.current_node.add_child(
                    TextNode(
                        parent=parsing_info.current_node,
                        cdata=True,
                        text=text[parsing_info.i: text[parsing_info.i:].find("]]>")]
                    )
                )
                parsing_info.move_index(len(text[parsing_info.i:].find("]]>") + len("]]>")))
            elif text[parsing_info.i:].find("<!--") == 0:
                comment_end_raw = text[parsing_info.i:].find("-->")
                if comment_end_raw == -1:
                    raise XMLParsingException("Not properly closed tag")
                comment_end = parsing_info.i + comment_end_raw
                parsing_info.current_node.add_child(CommentNode(content=text[parsing_info.i + 4: comment_end].strip()))
                parsing_info.move_index(len(f"{text[parsing_info.i: comment_end]}-->"))
            elif text[parsing_info.i + 1] == "/":
                name = text[parsing_info.i + 2: parsing_info.i + 2 + text[parsing_info.i + 2:].find(">")]
                if parsing_info.current_node.get_name() == name:
                    parsing_info.current_node = parsing_info.current_node.parent
                    parsing_info.move_index(len(f"</{name}>"))
                else:
                    print(text[parsing_info.i + 1:])
                    raise XMLParsingException()
            else:
                parsing_info.state = STATES.read_node_name
                parsing_info = self.create_new_node(text=text, parsing_info=parsing_info)

        else:
            parsing_info.move_index()

        return text, parsing_info

    def create_new_node(self, text: str, parsing_info) -> XmlParsingInfo:
        if text[parsing_info.i + 1] == "?":
            parsing_info.move_index(2)
            n = ProcessingNode(parent=parsing_info.current_node)
            parsing_info.current_node.add_child(n)
            parsing_info.current_node = n
            return parsing_info
        elif text[parsing_info.i + 1] == "!":
            parsing_info.move_index(2)
            n = DirectiveNode(parent=parsing_info.current_node)
            parsing_info.current_node.add_child(n)
            parsing_info.current_node = n
            return parsing_info
        else:
            parsing_info.move_index()
            n = Node(parent=parsing_info.current_node)
            parsing_info.current_node.add_child(n)
            parsing_info.current_node = n
            parsing_info.current_node.children = []  # there was some weirdness, TODO investigate later
            parsing_info.current_node.attributes = {}
            parsing_info.state = STATES.read_node_name
            return parsing_info

    def parse_ending_tag_character(self, text: str, parsing_info: XmlParsingInfo) -> (str, XmlParsingInfo):
        if parsing_info.state == STATES.looking_for_attribute:
            if (text[parsing_info.i - 1] == "/" or not parsing_info.current_node.is_paired()) \
                    and type(parsing_info.current_node) is not RootNode:
                parsing_info.current_node.paired_tag = False
                parsing_info.current_node = parsing_info.current_node.parent

            parsing_info.state = STATES.looking_for_child_nodes

            if parsing_info.current_node.get_name().lower() != 'script':
                parsing_info.move_index()
            else:
                end_script = parsing_info.i + text[parsing_info.i + 1:].find("</script>")
                if end_script < parsing_info.i:
                    raise XMLParsingException("Attribute not ended")
                parsing_info.current_node.add_child(
                    TextNode(
                        content=text[parsing_info.i + 1:end_script]
                    )
                )
                parsing_info.move_index(len(f"{text[parsing_info.i:end_script]}</script>") + 1)
                parsing_info.current_node = parsing_info.current_node.parent
        else:
            parsing_info.move_index()
        return text, parsing_info

    def parse_character(self, text: str, parsing_info: XmlParsingInfo) -> (str, XmlParsingInfo):
        if parsing_info.state == STATES.read_node_name:
            name_end = positive_min(
                text[parsing_info.i:].find(" "),
                text[parsing_info.i:].find(">"),
                text[parsing_info.i:].find("\n")
            ) + parsing_info.i
            parsing_info.current_node.set_name(text[parsing_info.i: name_end])
            parsing_info.move_index(len(text[parsing_info.i: name_end]))
            parsing_info.state = STATES.looking_for_attribute
        elif parsing_info.state == STATES.looking_for_attribute:
            if text[parsing_info.i].isalnum():
                attr_divider = text[parsing_info.i:].find("=")
                tag_end = text[parsing_info.i:].find(">")
                if tag_end == -1:
                    raise XMLParsingException("Not properly closed tag")
                if ((attr_divider > tag_end) and (tag_end >= 0)) or (attr_divider < 0):
                    name = text[parsing_info.i: parsing_info.i + positive_min(
                        text[parsing_info.i:].find(">"),
                        text[parsing_info.i:].find(" "),
                        text[parsing_info.i:].find("/>"),
                    )]
                    attribute_value = ""
                    parsing_info.move_index(len(f"{name}"))
                else:
                    name = text[parsing_info.i: parsing_info.i + text[parsing_info.i:].find("=")]
                    attribute_value = self.get_attribute_value(text=text, parsing_info=parsing_info)
                    parsing_info.move_index(len(f"{name}='{attribute_value}'"))
                parsing_info.current_node.attributes[name] = attribute_value
            else:
                parsing_info.move_index()
        elif parsing_info.state == STATES.looking_for_child_nodes:
            text_end_raw = text[parsing_info.i:].find("<")
            if text_end_raw == -1:
                raise XMLParsingException("Not properly closed tag")
            text_end = parsing_info.i + text_end_raw
            parsing_info.current_node.add_child(TextNode(content=text[parsing_info.i: text_end]))
            parsing_info.move_index(len(text[parsing_info.i: text_end]))
        else:
            parsing_info.move_index()
        return text, parsing_info

    def get_attribute_value(self, text: str, parsing_info: XmlParsingInfo) -> (str):
        ending_char = text[parsing_info.i + text[parsing_info.i:].find("=") + 1] if text[parsing_info.i + text[parsing_info.i:].find("=") + 1] in {"\"", "'"} else -1
        if ending_char == -1:
            raise XMLParsingException("Attribute not ended")
        j = text[parsing_info.i:].find("=") + parsing_info.i + len(f"{ending_char}")

        while j < len(text):
            j += 1
            if text[j] == ending_char and text[j - 1] != "\\":
                return text[text[parsing_info.i:].find("=") + parsing_info.i + len(f"{ending_char}") + 1: j]

File: src/main/scala/com/sloth/templating/XMLParser/XMLParser.scala
package com.sloth.templating.XMLParser

import com.sloth.templating.exceptions.{EmptyTemplateException, XMLParsingException}
import com.sloth.templating.nodes._
import com.sloth.utilities.Utilities

import scala.collection.mutable.ListBuffer
import scala.util.control.Breaks.break

/**
 * Class responsible for parsing XML into Node tree
 *
 * @param template A template to be parsed
 */
class XMLParser(template: String, rootNode: RootNode = new RootNode()) {
  val parsingInfo = new XMLParsingInfo(rootNode = rootNode, currentNode = rootNode)

  /**
   * Main function to be interacted with. Parses template into nodes.
   *
   * @return A new containing root node of the tree
   */
  def parse: Node = {
    if (this.template.isEmpty) {
      throw new EmptyTemplateException("template cannot be empty")
    }

    while (parsingInfo.idx < this.template.length) {
      this.template.charAt(parsingInfo.idx).toString match {
        case "<" => this.parseStartingTagCharacter()
        case ">" => this.parseEndingTagCharacter()
        case " " | "\n" | "\t" => parsingInfo.moveIndex()
        case _ => this.parseCharacter()
      }
    }

    new Node(name = "abc", children = ListBuffer(), argPairedTag = true)
  }

  /**
   * This function is called when "<" is encountered
   */
  def parseStartingTagCharacter(): Unit = {
    if (this.template(this.parsingInfo.idx + 1) == ' ') {
      // TODO test this, just in case
      this.parsingInfo.moveIndex()
    } else if (this.parsingInfo.state == ParsingStates.NEW_PAGE) {
      this.startingCharOnNewPage()
    } else if (this.parsingInfo.state == ParsingStates.LOOKING_FOR_CHILD_NODES) {
      this.startingCharLookingForChildNodes()
    } else {
      this.parsingInfo.moveIndex()
    }
  }

  /**
   * Creates a new node
   *
   * @return newly created Node
   */
  def createNewNode(): Node = {
    if (this.template(this.parsingInfo.idx + 1) == '?') {
      val n = new ProcessingNode(parent = this.parsingInfo.currentNode)
      this.parsingInfo.currentNode.addChild(n)
      this.parsingInfo.currentNode = n
      this.parsingInfo.moveIndex(2)
      this.parsingInfo.state = ParsingStates.READ_NODE_NAME
      n
    } else if (this.template(this.parsingInfo.idx + 1) == '!') {
      val n = new DirectiveNode(parent = this.parsingInfo.currentNode)
      this.parsingInfo.currentNode.addChild(n)
      this.parsingInfo.currentNode = n
      this.parsingInfo.moveIndex(2)
      this.parsingInfo.state = ParsingStates.READ_NODE_NAME
      n
    } else {
      val n = new Node(parent = this.parsingInfo.currentNode, name = "")
      this.parsingInfo.currentNode.addChild(n)
      this.parsingInfo.currentNode = n
      if (n.children.nonEmpty) {
        /*
        This is here just in case. There was an issue with Python's runtime because it did NOT create
        Nodes properly and list of children had to be created manually or the children would leak to
        unrelated objects.
        */
        throw new Error("Memory is leaking here again")
      }
      this.parsingInfo.state = ParsingStates.READ_NODE_NAME
      this.parsingInfo.moveIndex()
      n
    }
  }

  /**
   * This function is called when state is ParsingStates.NEW_PAGE and "<" was encountered.
   */
  def startingCharOnNewPage(): Unit = {
    if (this.template.substring(this.parsingInfo.idx).indexOf("<?xml") == 0) {
      parsingInfo.moveIndex("<?xml".length)
      parsingInfo.state = ParsingStates.LOOKING_FOR_ATTRIBUTE
    } else if (this.template.substring(this.parsingInfo.idx).indexOf("<!DOCTYPE") == 0) {
      this.parsingInfo.currentNode match {
        case node: RootNode => {
          node.html = true
          this.parsingInfo.currentNode.setAttribute(
            "doctype",
            this.template.substring(
              this.parsingInfo.idx + "<!DOCTYPE ".length,
              this.template.indexOf(">")
            ))
          this.parsingInfo.state = ParsingStates.LOOKING_FOR_CHILD_NODES
          this.parsingInfo.moveIndex(this.template.indexOf(">"))
        }
      }
    } else {
      this.parsingInfo.currentNode = this.createNewNode()
      this.parsingInfo.state = ParsingStates.READ_NODE_NAME
    }
  }

  /**
   * This function is called when state is ParsingStates.LOOKING_FOR_CHILD_NODES and "<" was encountered.
   */
  def startingCharLookingForChildNodes(): Unit = {
    if (this.template.substring(this.parsingInfo.idx).indexOf("<![CDATA[") == 0) {
      this.parsingInfo.moveIndex("<![CDATA[".length)
      val tn = new TextNode(
        parent = this.parsingInfo.currentNode,
        cdata = true,
        content = this.template.substring(
          this.parsingInfo.idx,
          this.template.substring(this.parsingInfo.idx).indexOf("]]>")
        )
      )
      this.parsingInfo.currentNode.addChild(tn)
      this.parsingInfo.moveIndex(s"${tn.content}]]>".length)
    } else if (this.template.substring(this.parsingInfo.idx).indexOf("<!--") == 0) {
      this.parsingInfo.moveIndex("<![CDATA[".length)
      val cn = new CommentNode(
        parent = this.parsingInfo.currentNode,
        content = this.template.substring(
          this.parsingInfo.idx,
          this.template.substring(this.parsingInfo.idx).indexOf("-->")
        )
      )
      this.parsingInfo.currentNode.addChild(cn)
      this.parsingInfo.moveIndex(s"${cn.content}-->".length)
    } else if (this.template(this.parsingInfo.idx + 1) == '/') {
      val name = this.template.substring(this.parsingInfo.idx + 2, this.parsingInfo.idx + 2 + this.template.substring(this.parsingInfo.idx + 2).indexOf(">"))
      if (this.parsingInfo.currentNode.name == name) {
        this.parsingInfo.currentNode = parsingInfo.currentNode.parent
        this.parsingInfo.moveIndex(s"</${name}>".length)
      } else {
        throw new XMLParsingException("Need to test this thoroughly")
      }
    } else {
      this.parsingInfo.state = ParsingStates.READ_NODE_NAME
      this.createNewNode()
    }
  }

  /**
   * Function that handles processing ">" character
   */
  def parseEndingTagCharacter(): Unit = {
    if (this.parsingInfo.state == ParsingStates.LOOKING_FOR_ATTRIBUTE) {
      if ((this.template.substring(this.parsingInfo.idx - 1) == "/" || !this.parsingInfo.currentNode.isPairedTag)
        && !this.parsingInfo.currentNode.isInstanceOf[RootNode]) {
        this.parsingInfo.currentNode.argPairedTag = false
        this.parsingInfo.currentNode = this.parsingInfo.currentNode.parent
      }
      this.parsingInfo.state = ParsingStates.LOOKING_FOR_CHILD_NODES

      if (this.parsingInfo.currentNode.name.toLowerCase != "script") {
        this.parsingInfo.moveIndex()
      } else {
        val endScriptPosition = this.parsingInfo.idx + this.template.substring(this.parsingInfo.idx).indexOf("</script>")
        if (endScriptPosition < this.parsingInfo.idx) {
          throw new XMLParsingException("Script attribute not ended")
        }
        val tn = new TextNode(
          cdata = false,
          content = this.template.substring(this.parsingInfo.idx + 1, endScriptPosition),
          parent = this.parsingInfo.currentNode
        )
        this.parsingInfo.currentNode.addChild(tn)
        this.parsingInfo.moveIndex(
          s"${this.template.substring(this.parsingInfo.idx, endScriptPosition)}</script>".length
        )
        this.parsingInfo.currentNode = this.parsingInfo.currentNode.parent
      }
    } else {
      this.parsingInfo.moveIndex()
    }
  }

  /**
   * Function which decides what to do when a character is encountered
   */
  def parseCharacter(): Unit = {
    parsingInfo.state match {
      case ParsingStates.READ_NODE_NAME => this.readNodeName()
      case ParsingStates.LOOKING_FOR_ATTRIBUTE => this.lookForAttribute()
      case ParsingStates.LOOKING_FOR_CHILD_NODES => this.lookingForChildTextNodes()
      case _ => this.parsingInfo.moveIndex()
    }
  }

  /**
   * Reads a name of the element and moves parser further
   */
  def readNodeName(): Unit = {
    val nameEnd = Utilities.positiveMinInt(List(
      this.template.substring(parsingInfo.idx).indexOf(" "),
      this.template.substring(parsingInfo.idx).indexOf(">"),
      this.template.substring(parsingInfo.idx).indexOf("/>"),
      this.template.substring(parsingInfo.idx).indexOf("\n")
    )) + parsingInfo.idx
    parsingInfo.currentNode = parsingInfo.currentNode.setName(this.template.substring(parsingInfo.idx, nameEnd))
    parsingInfo.moveIndex(this.template.substring(parsingInfo.idx, nameEnd).length)
    parsingInfo.state = ParsingStates.LOOKING_FOR_ATTRIBUTE
  }

  /**
   * Function dealing with attributes in current node
   */
  def lookForAttribute(): Unit = {
    if (this.template.charAt(parsingInfo.idx).isLetterOrDigit) {
      val attrDivider = this.template.substring(parsingInfo.idx).indexOf("=")
      val tagEnd = this.template.substring(parsingInfo.idx).indexOf(">")

      if (tagEnd == -1) {
        throw new XMLParsingException("Not properly closed tag")
      }
      if (attrDivider > tagEnd || attrDivider < 0) {
        val attrName = this.template.substring(parsingInfo.idx, parsingInfo.idx + Utilities.positiveMinInt(List(
          this.template.substring(parsingInfo.idx).indexOf(" "),
          this.template.substring(parsingInfo.idx).indexOf(">"),
          this.template.substring(parsingInfo.idx).indexOf("/>"),
        )))
        val attrValue = ""
        parsingInfo.currentNode.setAttribute(attrName, attrValue)
        parsingInfo.moveIndex(attrName.length)
      } else {
        val equalsPosition = parsingInfo.idx + this.template.substring(parsingInfo.idx).indexOf("=")
        val attrName = this.template.substring(parsingInfo.idx, equalsPosition)
        val attrValQuoteChar = this.template.charAt(parsingInfo.idx + this.template.substring(parsingInfo.idx).indexOf("=") + 1)
        if (attrValQuoteChar.toString != "\"" && attrValQuoteChar.toString != "'") {
          throw new XMLParsingException("Attribute is not in quotes")
        }
        var attrValEndCharPosition: Int = parsingInfo.idx + this.template.substring(parsingInfo.idx).indexOf("=") + 1
        val attrValue: String = {
          var result = ""
          var resultChanged = false
          do {
            attrValEndCharPosition += 1
            if (this.template.charAt(attrValEndCharPosition) == attrValQuoteChar && this.template.charAt(attrValEndCharPosition - 1).toString != "\\") {
              result = this.template.substring(equalsPosition + 2, attrValEndCharPosition)
              resultChanged = true
            }
          } while ((attrValEndCharPosition < this.template.length - 1) && !(this.template.charAt(attrValEndCharPosition) == attrValQuoteChar && this.template.charAt(attrValEndCharPosition - 1).toString != "\\"))
          if (!resultChanged) {
            throw new XMLParsingException("XML file has unpaired quotes")
          }
          result
        }
        parsingInfo.currentNode.setAttribute(attrName, attrValue)
        parsingInfo.moveIndex(f"$attrName='$attrValue'".length + 1)
      }
    } else {
      parsingInfo.moveIndex()
    }
  }

  /**
   * Function processes adding child text node
   */
  def lookingForChildTextNodes(): Unit = {
    val textEndRaw = this.template.substring(parsingInfo.idx).indexOf("<")
    if (textEndRaw == -1) {
      throw new XMLParsingException("Not properly closed tag")
    }
    val textEnd = parsingInfo.idx + textEndRaw
    parsingInfo.currentNode.addChild(new TextNode(content = this.template.substring(parsingInfo.idx, textEnd), parent = parsingInfo.currentNode))
    parsingInfo.moveIndex(this.template.substring(parsingInfo.idx, textEnd).length)
  }
}
